### PR TITLE
State Install without project - Review

### PR DESCRIPTION
**Pull request by Naatan**
https://www.pivotaltracker.com/story/show/178223554

### CODE REVIEW

File: cmd/state/internal/cmdtree/cmdtree.go [hunk 1]
```diff
@@ -319,15 +311,3 @@ func (a *addCmdAs) deprecatedAlias(aliased *captain.Command, name string) {
 	a.parent.AddChildren(cmd)
 }
 
-func interceptAddHeadlessNotify(out output.Outputer, pj *project.Project) captain.InterceptFunc {
```
Comment (Naatan) [inline cmd/state/internal/cmdtree/cmdtree.go:322]: Not needed anymore due to the local env management changes we're making.

File: internal/runners/packages/install.go [hunk 2]
```diff
@@ -16,26 +12,16 @@ type InstallRunParams struct {
 
 // Install manages the installing execution context.
 type Install struct {
-	cfg  configurable
-	out  output.Outputer
-	proj *project.Project
-	prompt.Prompter
-	auth *authentication.Auth
+	prime primeable
```
Comment (Naatan) [inline internal/runners/packages/install.go:15]: We should normally avoid doing this, but in this case passing `prime` along to the package operation saves us a lot of useless boilerplate.

File: internal/runners/packages/packages.go [hunk 3]
```diff
@@ -43,23 +42,22 @@ type configurable interface {
...
-			return locale.WrapError(err, "err_install_get_langauge", "Could not get language for package: {{.V0}}", packageName)
+			return locale.WrapError(err, "err_package_get_project", "Could not get project from path")
 		}
+		defer func() {
+			if rerr != nil {
+				if err := os.Remove(pj.Source().Path()); err != nil {
+					logging.Error("could not remove temporary project file: %s", errs.JoinMessage(err))
+				}
+			}
+		}()
```
Comment (Naatan) [inline internal/runners/packages/packages.go:60]: This ensure we remove the as.yaml we created if the operation failed.

File: pkg/platform/model/inventory.go [hunk 4]
```diff
@@ -83,32 +93,56 @@ func FetchAuthors(ingredID, ingredVersionID *strfmt.UUID) (Authors, error) {
 	return results.Payload.Authors, nil
 }
 
-func searchIngredientsNamespace(limit int, ns Namespace, name string) ([]*IngredientAndVersion, error) {
-	lim := int64(limit)
+func searchIngredientsNamespace(ns Namespace, name string, includeVersions bool) ([]*IngredientAndVersion, error) {
```
Comment (Naatan) [inline pkg/platform/model/inventory.go:96]: I modified this function to make use of pagination as well as to add a param allowing you to include version variants of the ingredients or just the base ingredient itself.

File: pkg/platform/model/recipe.go [hunk 5]
```diff
@@ -193,6 +174,24 @@ func FetchRecipe(commitID strfmt.UUID, owner, project string, hostPlatform *stri
 	return nil, locale.NewInputError("err_recipe_not_found")
 }
 
+func resolveSolverError(err error) error {
```
Comment (Naatan) [inline pkg/platform/model/recipe.go:177]: Proper solver errors

File: internal/runners/packages/packages.go [hunk 6]
```diff
@@ -154,22 +161,80 @@ func executePackageOperation(pj *project.Project, cfg configurable, out output.O
...
+		}
+		return "", ns, locale.WrapError(err, "err_pkgop_search",
+			"Could not match {{.V0}}. Did you mean:\n\n{{.V1}}", packageName, strings.Join(suggestions, "\n"))
+	}
+
+	//
+	choices := []string{}
+	values := map[string][]string{}
+	for _, i := range ingredients {
+		language := model.LanguageFromNamespace(*i.Ingredient.PrimaryNamespace)
```
Comment (MDrakos) [inline internal/runners/packages/packages.go:190]: While it's unlikely that this value is the empty string, not checking it here can lead to some odd display for the user and invalid namespaces. We could easily check if it's empty and then log and continue.

File: internal/runners/packages/packages.go [hunk 7]
```diff
@@ -154,22 +161,80 @@ func executePackageOperation(pj *project.Project, cfg configurable, out output.O
...
+		// Generate ingredient choices to present to the user
+		name := fmt.Sprintf("%s (%s)", *i.Ingredient.Name, language)
+		choices = append(choices, name)
+		values[name] = []string{*i.Ingredient.Name, language}
+	}
+
+	// Prompt the user with the ingredient choices
+	choice, err := prompt.Select(
+		locale.Tl("prompt_pkgop_ingredient", "Multiple Matches"),
+		locale.Tl("prompt_pkgop_ingredient_msg", "Multiple ingredients matched your query, which one would you like to use?"),
```
Comment (MDrakos) [inline internal/runners/packages/packages.go:207]: Is `ingredient` a term we expect the user to understand? I'm not sure if we're exposing them to our kitchen metaphors anywhere else.

**Review by MDrakos — APPROVED** [review]

File: internal/runners/packages/packages.go (hunk 6 above)
Comment (Naatan) [reply internal/runners/packages/packages.go:190]: This won't ever be nil, that's a swagger oddity. But it is also guarded against via the `SearchIngredientsStrict` function.

**Pull request merged**